src/main/java/com/matt/forgehax/mods/StepMod.java

```diff
@@ -29,27 +32,80 @@
           .description("how high you can step")
           .defaultTo(1.2f)
           .min(0f)
+          .success(__ -> {
+            if (isEnabled()) {
+              EntityPlayer player = getLocalPlayer();
+              if (player != null) updateStepHeight(player);
+            }
+          })
+          .build();
+
+  public final Setting<Boolean> unstep =
+      getCommandStub()
+          .builders()
+          .<Boolean>newSettingBuilder()
+          .name("unstep")
+          .description("step down instead of falling")
+          .defaultTo(false)
           .build();
 
   public StepMod() {
     super(Category.PLAYER, "Step", false, "Step up blocks");
   }
 
+  @Override
+  protected void onEnabled() {
+    EntityPlayer player = getLocalPlayer();
+    if (player != null) wasOnGround = player.onGround;
+  }
+
   @Override
   public void onDisabled() {
-    if (getLocalPlayer() != null) {
-      getLocalPlayer().stepHeight = DEFAULT_STEP_HEIGHT;
+    EntityPlayer player = getLocalPlayer();
+    if (player != null) player.stepHeight = DEFAULT_STEP_HEIGHT;
+  }
+
+  void updateStepHeight(EntityPlayer player) {
+    if (player.onGround) {
+      player.stepHeight = stepHeight.get();
+    } else {
+      player.stepHeight = DEFAULT_STEP_HEIGHT;
+    }
+  }
+
+  public boolean wasOnGround = false;
+
+  public void unstep(EntityPlayer player) {
+    if (!MC.isCallingFromMinecraftThread())
+      throw new IllegalStateException("Call this using MC.addScheduledTask");
+
+    AxisAlignedBB range = player.getEntityBoundingBox().expand(0, -stepHeight.get(), 0).contract(0, player.height, 0);
+
+    if (!player.world.collidesWithAnyBlock(range)) return;
+
+    List<AxisAlignedBB> collisionBoxes = player.world.getCollisionBoxes(player, range);
+    AtomicReference<Double> newY = new AtomicReference<>(0D);
```

Comment:
Why `AtomicReference`?

Reply:
My IDE wanted me to use this instead of a regular double because I updated it in a lambda. I know that iteration order is guaranteed and nothing is executed in parallel.